### DISCUSSION TITLE
Camera virtual scene navigation

### DISCUSSION THREAD

**Discussion by jp833** [appium/appium#16048] [Q&A] [answered]
Is it possible to navigate the camera virtual scene? We are trying to automate our augmented reality react-native native app.
PressKeyCode or LongPressKeyCode do not seem to work for Alt+WASDZE.

**Comment by KazuCocoa** [comment, marked as answer]
Is it https://developers.google.com/ar/develop/java/emulator#add_augmented_images_to_the_scene ...?
It seems https://developers.google.com/ar/develop/java/emulator#move_the_virtual_camera expects to get keys as its host OS. (Not Android emulator level's `adb` command etc.)

If your client and appium server/the android emulator work on the same macOS host, I guess AppleScript will help. (not via appium.)
https://stackoverflow.com/questions/13818889/applescript-to-invoke-alt-keypress-on-cmdtap-keypress-to-switch-to-hidden-windo

**Reply by jp833** [reply]
Yes it is https://developers.google.com/ar/develop/java/emulator#add_augmented_images_to_the_scene. AppleScript looks promising. Thanks. Will give it a try.

**Reply by jp833** [reply]
AppleScript worked. To automate - I invoke the script from Appium using 

```
string cmd = "osascript /Users/jyotipandey/Desktop//Automation.scpt";
            var escapedArgs = cmd.Replace("\"", "\\\"");

            var process = new Process()
            {
                StartInfo = new ProcessStartInfo
                {
                    FileName = "/bin/bash",
                    Arguments = $"-c \"{escapedArgs}\"",
                    RedirectStandardOutput = true,
                    UseShellExecute = false,
                    CreateNoWindow = true,
                }
            };

            var result = process.Start();
```

And my script looks like this -- 

```
tell application "System Events"
	
	try
		key down option
		
		repeat 500 times
			tell application "System Events"
				key down "a"
			end tell
		end repeat
		key up "a"
		delay 5
		repeat 500 times
			tell application "System Events"
				key down "d"
			end tell
		end repeat
		key up "d"
		delay 5
		
		repeat 500 times
			tell application "System Events"
				key down "w"
			end tell
		end repeat
		
		key up "w"
		delay 5
		
		repeat 500 times
			tell application "System Events"
				key down "s"
			end tell
		end repeat
		key up "s"
		delay 5
	end try
	
	key up option
end tell

```
